vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1259 @@
+        }
+
+        for (auto& pip : pips_) {
+            auto& key = pip.first;
+            auto& value = pip.second;
+            std::tuple<int, int, int, bool> meta;
+            int switch_id, tile, wire0, wire1;
+            bool forward;
+            std::tie(switch_id, meta) = value;
+            std::tie(tile, wire0, wire1, forward) = meta;
+            auto& r1 = redirect_[std::tuple<int, loc_>(std::get<0>(key), tile_to_loc_[tile])];
+            auto& r2 = redirect_[std::tuple<int, loc_>(std::get<1>(key), tile_to_loc_[tile])];
+            VTR_LOG("Switch_type: %d, %s.%s %s.%s: forward:%s\n",
+                    switch_id, str(tile).c_str(), str(wire0).c_str(), str(tile).c_str(), str(wire1).c_str(), forward ? "yes" : "no");
+            VTR_LOG("Edge: CHAN%c X:%d Y:%d idx:%d -> CHAN%c X:%d Y:%d idx%d\n",
+                    std::get<1>(r1) == e_rr_type::CHANX ? 'X' : 'Y',
+                    std::get<0>(std::get<0>(r1)),
+                    std::get<1>(std::get<0>(r1)),
+                    std::get<2>(r1),
+                    std::get<1>(r2) == e_rr_type::CHANX ? 'X' : 'Y',
+                    std::get<0>(std::get<0>(r2)),
+                    std::get<1>(std::get<0>(r2)),
+                    std::get<2>(r2));
+        }
+        VTR_LOG("RR_Nodes\n");
+        for (auto i : loc_type_idx_to_rr_idx_) {
+            loc_ loc;
+            e_rr_type type;
+            int idx;
+            std::tie(loc, type, idx) = i.first;
+            VTR_LOG("\t X:%d Y:%d type:%d idx:%d->rr_node:%d\n", std::get<0>(loc), std::get<1>(loc), type, idx, i.second);
+        }
+    }
+
+    /*
+     * Fill device_ctx rr_switch_inf structure and store id of each PIP type for future use
+     */
+
+    void create_switch_inf() {
+        std::set<std::tuple<int /*timing*/, bool /*buffered*/>> seen;
+        for (const auto& tile : ar_.getTileTypeList()) {
+            for (const auto& pip : tile.getPips()) {
+                if (pip.isPseudoCells())
+                    continue;
+                std::tuple<int, bool> t1{pip.getTiming(), pip.getBuffered21()};
+                seen.insert(t1);
+                if (!pip.getDirectional()) {
+                    std::tuple<int, bool> t2{pip.getTiming(), pip.getBuffered20()};
+                    seen.insert(t2);
+                }
+            }
+        }
+        device_ctx_.rr_switch_inf.reserve(seen.size() + 1);
+        int id = 2;
+        make_room_in_vector(&device_ctx_.rr_switch_inf, 0);
+        device_ctx_.rr_switch_inf[0].R = 0;
+        device_ctx_.rr_switch_inf[0].Cin = 0;
+        device_ctx_.rr_switch_inf[0].Cout = 0;
+        device_ctx_.rr_switch_inf[0].Cinternal = 0;
+        device_ctx_.rr_switch_inf[0].Tdel = 0;
+        device_ctx_.rr_switch_inf[0].buf_size = 0;
+        device_ctx_.rr_switch_inf[0].mux_trans_size = 0;
+        device_ctx_.rr_switch_inf[0].name = "short";
+        device_ctx_.rr_switch_inf[0].set_type(SwitchType::SHORT);
```

Comment:
We may consider having a utility function to generate an empty switch to which we just pass the name and type, and optionally the rest of the values.